src/upstream_service.cpp

```diff
@@ -0,0 +1,294 @@
+  // NaNs serialize as an empty string.
+  if (std::isnan(value)) {
+    return;
+  }
+
+  // `printf`-style formatting rounds if necessary, but we need it to round
+  // _up_.  So, use the ceiling function to get the upward rounding behavior,
+  // and then format the number.  `double` has enough precision that I don't
+  // think this is any different than formatting the number as decimal and
+  // then rounding up based on that.
+  const double decimal_shift = 1e4;
+  const double rounded = std::ceil(value * decimal_shift) / decimal_shift;
+
+  const char* const format = "%1.4f";
+
+  const int formatted_strlen = std::snprintf(nullptr, 0, format, rounded);
+  // The +1 is for the null terminator that we will later remove.
+  const int buffer_size = formatted_strlen + 1;
+  assert(formatted_strlen > 0);
+  destination.resize(destination.size() + buffer_size);
+
+  const int rcode =
+      std::snprintf(&*(destination.end() - buffer_size), buffer_size, format, rounded);
+  (void)rcode;  // Some `assert` macros do this, some don't.
+  assert(rcode == formatted_strlen);
+
+  // Remove the null terminator added by `snprintf`.
+  destination.pop_back();
+}
+
+bool operator==(const UpstreamService& left, const UpstreamService& right) {
+  // clang-format off
+  return left.service_name == right.service_name
+      && left.sampling_priority == right.sampling_priority
+      && left.sampling_mechanism == right.sampling_mechanism
+      // Two things to watch out for with `.sampling_rate`:
+      //
+      // 1. NaN is used as the N/A value, so we have to work around how NaN !=
+      //    NaN.
+      // 2. I want `original == deserialize(serialize(original))`.  This
+      //    property is handy in the unit test.  It's complicated by the
+      //    limited precision used in the formatting of `.sampling_rate`.  
+      //
+      // We handle both cases by comparing serialized versions of `.sampling_rate`.
+      && serializeSamplingRate(left.sampling_rate) == serializeSamplingRate(right.sampling_rate)
+      && left.unknown_fields == right.unknown_fields;
+  // clang-format on
+}
+
+std::vector<UpstreamService> deserializeUpstreamServices(ot::string_view text) {
+  std::vector<UpstreamService> result;
+
+  auto iter = text.begin();
+  const auto end = text.end();
+  if (iter == end) {
+    // An empty string means no upstream services.
+    return result;
+  }
+
+  decltype(iter) next;
+  do {
+    next = std::find(iter, end, ';');
+    result.push_back(deserialize(range(iter, next)));
+    iter = next + 1;
+  } while (next != end);
+
+  return result;
+}
+
+std::string serializeUpstreamServices(const std::vector<UpstreamService>& upstream_services) {
+  std::string result;
+  if (upstream_services.empty()) {
+    return result;
+  }
+
+  auto iter = upstream_services.begin();
+  serialize(result, *iter);
+  for (++iter; iter != upstream_services.end(); ++iter) {
+    result += ';';
+    serialize(result, *iter);
+  }
```

Comment:
Caleb: Another way of doing this... just the loop, delimiter first, but start the delimiter as empty and change it at the end of the loop body.